File: apps/android/app/src/main/java/com/interviewagent/data/InterviewApiClient.kt
package com.interviewagent.data

import android.util.Base64
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

class InterviewApiClient(
    private val baseUrl: String,
    private var token: String?
) {
    fun devLogin(userId: String = "android-dev-user"): AuthTokenResponse {
        val body = JSONObject()
            .put("user_id", userId)
            .put("display_name", "Android 开发用户")
            .put("platform", "android")
        val json = JSONObject(request("/auth/dev-login", "POST", body.toString()))
        val accessToken = json.getString("access_token")
        token = accessToken
        return AuthTokenResponse(
            accessToken = accessToken,
            expiresAt = json.optLong("expires_at"),
            tenantId = json.optString("tenant_id"),
            userId = json.optString("user_id"),
            platform = json.optString("platform"),
            displayName = json.optString("display_name"),
            trialUsesRemaining = json.optInt("trial_uses_remaining"),
            creditBalance = json.optString("credit_balance", "0")
        )
    }

    fun logout() {
        token = null
    }

    fun account(): AccountResponse {
        return parseAccount(request("/account"))
    }

    fun settings(): UserSettingsResponse {
        val json = JSONObject(request("/settings"))
        return UserSettingsResponse(defaultInterviewMode = json.optString("default_interview_mode", "interviewer"))
    }

    fun updateDefaultInterviewMode(mode: String): UserSettingsResponse {
        val body = JSONObject().put("default_interview_mode", mode)
        val json = JSONObject(request("/settings", "PUT", body.toString()))
        return UserSettingsResponse(defaultInterviewMode = json.optString("default_interview_mode", mode))
    }

    fun recharge(amountCredits: String): AccountResponse {
        val body = JSONObject()
            .put("amount_credits", amountCredits)
            .put("payment_provider", "android-mock")
            .put("external_order_id", "android-${System.currentTimeMillis()}")
        return parseAccount(request("/account/recharge", "POST", body.toString()))
    }

    fun health(): String {
        val payload = request("/health")
        return JSONObject(payload).optString("status", "unknown")
    }

    fun listIndustries(targetRole: String = "AI 应用工程师"): List<IndustryOption> {
        val encoded = URLEncoder.encode(targetRole, StandardCharsets.UTF_8.name())
        val payload = request("/metadata/industries?target_role=$encoded")
        val array = JSONArray(payload)
        return (0 until array.length()).map { index ->
            val item = array.getJSONObject(index)
            IndustryOption(
                value = item.getString("value"),
                label = item.getString("label"),
                description = item.optString("description"),
                productionSignals = item.optJSONArray("production_signals").toStringList(),
                recommendedFocusAreas = item.optJSONArray("recommended_focus_areas").toStringList()
            )
        }
    }

    fun listPracticeCategories(): List<PracticeCategory> {
        val array = JSONArray(request("/practice/categories"))
        return (0 until array.length()).map { index ->
            val item = array.getJSONObject(index)
            PracticeCategory(
                value = item.optString("value"),
                label = item.optString("label"),
                description = item.optString("description"),
                subjects = item.optJSONArray("subjects").toStringList()
            )
        }
    }

    fun listPracticeQuestions(category: String, limit: Int = 50): PracticeQuestionListResponse {
        val encoded = URLEncoder.encode(category, StandardCharsets.UTF_8.name())
        val json = JSONObject(request("/practice/questions?category=$encoded&limit=$limit"))
        val items = json.optJSONArray("items")
        return PracticeQuestionListResponse(
            items = (0 until (items?.length() ?: 0)).map { index -> parsePracticeQuestion(items?.getJSONObject(index) ?: JSONObject()) },
            total = json.optInt("total"),
            limit = json.optInt("limit"),
            offset = json.optInt("offset")
        )
    }

    fun seedPracticeQuestions(): ImportResultResponse {
        val json = JSONObject(request("/practice/questions/seed", "POST"))
        return ImportResultResponse(
            created = json.optInt("created"),
            updated = json.optInt("updated"),
            total = json.optInt("total")
        )
    }

    fun submitPracticeAttempt(questionId: String, answer: String, elapsedSeconds: Int): PracticeAttemptResponse {
        val body = JSONObject()
            .put("question_id", questionId)
            .put("answer", answer)
            .put("elapsed_seconds", elapsedSeconds)
        val json = JSONObject(request("/practice/attempt", "POST", body.toString()))
        return PracticeAttemptResponse(
            questionId = json.optString("question_id"),
            correct = if (json.isNull("correct")) null else json.optBoolean("correct"),
            score = json.optInt("score"),
            feedback = json.optString("feedback"),
            referenceAnswer = json.optString("reference_answer"),
            explanation = json.optString("explanation"),
            suggestions = json.optJSONArray("suggestions").toStringList(),
            elapsedSeconds = if (json.isNull("elapsed_seconds")) null else json.optInt("elapsed_seconds")
        )
    }

    fun createSession(request: CreateSessionRequest): ChatResponse {
        val body = JSONObject()
            .put("offline", request.offline)
            .put("web_search", request.webSearch)
            .put("mode", request.mode)
            .put("industry", request.industry)
            .put("target_role", request.targetRole)
            .put("seniority", request.seniority)
            .put("interview_goal", request.interviewGoal)
            .put("focus_areas", JSONArray(request.focusAreas))
        if (!request.resumeId.isNullOrBlank()) {
            body.put("resume_id", request.resumeId)
        }
        return parseChatResponse(request("/sessions", "POST", body.toString()))
    }

    fun listResumes(): List<ResumeRecord> {
        val array = JSONArray(request("/resumes"))
        return (0 until array.length()).map { index -> parseResume(array.getJSONObject(index)) }
    }

    fun importResume(filename: String, text: String): ResumeRecord {
        val encoded = Base64.encodeToString(text.toByteArray(StandardCharsets.UTF_8), Base64.NO_WRAP)
        val body = JSONObject()
            .put("filename", filename)
            .put("content_base64", encoded)
        return parseResume(JSONObject(request("/resumes", "POST", body.toString())))
    }

    fun deleteResume(id: String): Boolean {
        return JSONObject(request("/resumes/$id", "DELETE")).optBoolean("deleted")
    }

    fun listSessions(limit: Int = 50): List<SessionSummary> {
        val array = JSONArray(request("/sessions?limit=$limit"))
        return (0 until array.length()).map { index ->
            val item = array.getJSONObject(index)
            SessionSummary(
                id = item.getString("id"),
                resumeId = item.optString("resume_id").ifBlank { null },
                mode = item.optString("mode"),
                industry = item.optString("industry"),
                candidateName = item.optString("candidate_name"),
                targetRole = item.optString("target_role"),
                seniority = item.optString("seniority"),
                status = item.optString("status"),
                createdAt = item.optString("created_at"),
                updatedAt = item.optString("updated_at")
            )
        }
    }

    fun getSession(id: String): SessionDetail {
        val json = JSONObject(request("/sessions/$id"))
        val turns = json.optJSONArray("turns")
        return SessionDetail(
            id = json.getString("id"),
            turns = (0 until (turns?.length() ?: 0)).map { index ->
                val item = turns?.getJSONObject(index) ?: JSONObject()
                TurnPayload(
                    interviewer = item.optString("interviewer").ifBlank { null },
                    candidate = item.optString("candidate").ifBlank { null }
                )
            }
        )
    }

    fun deleteSession(id: String): Boolean {
        return JSONObject(request("/sessions/$id", "DELETE")).optBoolean("deleted")
    }

    fun sendMessage(sessionId: String, message: String): ChatResponse {
        val body = JSONObject().put("message", message)
        return parseChatResponse(request("/sessions/$sessionId/messages", "POST", body.toString()))
    }

    fun streamMessage(sessionId: String, message: String): List<StreamEvent> {
        val body = JSONObject().put("message", message)
        return parseSse(request("/sessions/$sessionId/stream", "POST", body.toString()))
    }

    private fun parseChatResponse(payload: String): ChatResponse {
        val json = JSONObject(payload)
        return ChatResponse(
            sessionId = json.getString("session_id"),
            message = json.getString("message"),
            completed = json.optBoolean("completed"),
            fallbackUsed = json.optBoolean("fallback_used"),
            guardrails = json.optJSONArray("guardrails").toStringList()
        )
    }

    private fun parseAccount(payload: String): AccountResponse {
        val json = JSONObject(payload)
        return AccountResponse(
            tenantId = json.optString("tenant_id"),
            userId = json.optString("user_id"),
            displayName = json.optString("display_name"),
            email = json.optString("email").ifBlank { null },
            platform = json.optString("platform"),
            trialUsesRemaining = json.optInt("trial_uses_remaining"),
            creditBalance = json.optString("credit_balance", "0"),
            creditBalanceMicros = json.optLong("credit_balance_micros")
        )
    }

    private fun parseResume(json: JSONObject): ResumeRecord {
        return ResumeRecord(
            id = json.getString("id"),
            filename = json.optString("filename"),
            fileType = json.optString("file_type"),
            summary = json.optString("summary"),
            text = json.optString("text"),
            truncated = json.optBoolean("truncated"),
            createdAt = json.optString("created_at"),
            updatedAt = json.optString("updated_at")
        )
    }

    private fun parsePracticeQuestion(json: JSONObject): PracticeQuestion {
        return PracticeQuestion(
            id = json.optString("id"),
            practiceCategory = json.optString("practice_category"),
            examYear = json.optInt("exam_year"),
            examName = json.optString("exam_name"),
            subject = json.optString("subject"),
            questionType = json.optString("question_type"),
            prompt = json.optString("prompt"),
            choices = json.optJSONArray("choices").toStringList(),
            answer = json.optString("answer").ifBlank { null },
            explanation = json.optString("explanation").ifBlank { null },
            difficulty = json.optString("difficulty")
        )
    }

    private fun request(path: String, method: String = "GET", body: String? = null): String {
        val connection = URL("$baseUrl$path").openConnection() as HttpURLConnection
        connection.requestMethod = method
        connection.setRequestProperty("Content-Type", "application/json")
        if (!token.isNullOrBlank()) {
            connection.setRequestProperty("Authorization", "Bearer $token")
        }
        if (body != null) {
            connection.doOutput = true
            connection.outputStream.use { stream ->
                stream.write(body.toByteArray(StandardCharsets.UTF_8))
            }
        }

        val status = connection.responseCode
        val stream = if (status in 200..299) connection.inputStream else connection.errorStream
        val text = stream?.let {
            BufferedReader(InputStreamReader(it, StandardCharsets.UTF_8)).use { reader ->
                reader.readText()
            }
        }.orEmpty()
        if (status !in 200..299) {
            val detail = runCatching { JSONObject(text).optString("detail") }.getOrDefault("")
            throw IllegalStateException(detail.ifBlank { "HTTP $status" })
        }
        return text
    }

    private fun parseSse(text: String): List<StreamEvent> {
        return text.split("\n\n")
            .map { it.trim() }
            .filter { it.isNotBlank() }
            .map { block ->
                val lines = block.lines()
                val event = lines.firstOrNull { it.startsWith("event:") }
                    ?.removePrefix("event:")
                    ?.trim()
                    ?: "message"
                val dataText = lines.firstOrNull { it.startsWith("data:") }
                    ?.removePrefix("data:")
                    ?.trim()
                    ?: "{}"
                StreamEvent(event = event, data = JSONObject(dataText).toMap())
            }
    }
}

private fun JSONArray?.toStringList(): List<String> {
    if (this == null) return emptyList()
    return (0 until length()).map { index -> optString(index) }
}

private fun JSONObject.toMap(): Map<String, Any?> {
    return keys().asSequence().associateWith { key ->
        when (val value = get(key)) {
            is JSONObject -> value.toMap()
            is JSONArray -> value.toStringList()
            else -> value
        }
    }
}

File: apps/ios/Sources/InterviewAgent/InterviewApiClient.swift
import Foundation

final class InterviewApiClient {
    private let baseURL: URL
    private var token: String?
    private let session: URLSession
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder

    private let onTokenChanged: (String?) -> Void

    private var clientVersion: String {
        Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "unknown"
    }

    init(baseURL: URL, token: String?, session: URLSession = .shared, onTokenChanged: @escaping (String?) -> Void = { _ in }) {
        self.baseURL = baseURL
        self.token = token
        self.session = session
        self.encoder = JSONEncoder()
        self.decoder = JSONDecoder()
        self.onTokenChanged = onTokenChanged
    }

    func health() async throws -> HealthResponse {
        try await request(path: "/health")
    }

    func login(email: String, password: String) async throws -> AuthTokenResponse {
        let response: AuthTokenResponse = try await request(path: "/auth/login", method: "POST", body: ["email": email.trimmingCharacters(in: .whitespacesAndNewlines), "password": password.trimmingCharacters(in: .whitespacesAndNewlines), "platform": "ios"])
        token = response.accessToken
        onTokenChanged(response.accessToken)
        return response
    }

    func register(email: String, password: String, displayName: String) async throws -> AuthTokenResponse {
        let response: AuthTokenResponse = try await request(path: "/auth/register", method: "POST", body: ["email": email.trimmingCharacters(in: .whitespacesAndNewlines), "password": password.trimmingCharacters(in: .whitespacesAndNewlines), "display_name": displayName.trimmingCharacters(in: .whitespacesAndNewlines), "platform": "ios"])
        token = response.accessToken
        onTokenChanged(response.accessToken)
        return response
    }

    func logout() {
        token = nil
        onTokenChanged(nil)
    }

    func learningToday() async throws -> LearningTodayResponse { try await request(path: "/learning/today") }

    func commandLearningTask(_ task: LearningTask, action: String) async throws -> LearningTask {
        let response: LearningCommandResponse = try await request(path: "/learning/tasks/\(task.id)/commands", method: "POST", body: LearningCommandRequest(action: action, expectedVersion: task.version), headers: ["Idempotency-Key": "ios-\(UUID().uuidString)"])
        return response.task
    }

    func listReviewPlans(limit: Int = 20, offset: Int = 0) async throws -> [ReviewPlan] { try await request(path: "/review-site/plans?limit=\(limit)&offset=\(offset)") }
    func generateReviewPlan(_ body: PlanGenerateRequest) async throws -> PlanGenerateResponse { try await request(path: "/review-site/planner/generate", method: "POST", body: body) }
    func checkin(planID: String, minutes: Int, note: String) async throws -> CheckinResponse { try await request(path: "/review-site/plans/\(planID)/checkin", method: "POST", body: CheckinRequest(elapsedMinutes: minutes, note: note)) }
    func listInterviewKits(limit: Int = 20, offset: Int = 0) async throws -> [InterviewKit] { try await request(path: "/interviewer-workspace/kits?limit=\(limit)&offset=\(offset)") }
    func createInterviewKit(_ body: InterviewKitRequest) async throws -> InterviewKit { try await request(path: "/interviewer-workspace/kits", method: "POST", body: body) }

    func account() async throws -> AccountResponse {
        try await request(path: "/account")
    }

    func getSettings() async throws -> UserSettingsResponse {
        try await request(path: "/settings")
    }

    func updateDefaultInterviewMode(_ mode: InterviewMode) async throws -> UserSettingsResponse {
        try await request(
            path: "/settings",
            method: "PUT",
            body: ["default_interview_mode": mode.rawValue]
        )
    }

    func recharge(amountCredits: String) async throws -> AccountResponse {
        try await request(
            path: "/account/recharge",
            method: "POST",
            body: [
                "amount_credits": amountCredits,
                "payment_provider": "ios-mock",
                "external_order_id": "ios-\(Int(Date().timeIntervalSince1970 * 1000))"
            ]
        )
    }

    func listIndustries(targetRole: String = "AI 应用工程师") async throws -> [IndustryOption] {
        let encoded = targetRole.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? targetRole
        return try await request(path: "/metadata/industries?target_role=\(encoded)")
    }

    func listPracticeCategories() async throws -> [PracticeCategory] {
        try await request(path: "/practice/categories")
    }

    func listPracticeQuestions(category: String, limit: Int = 20, offset: Int = 0) async throws -> PracticeQuestionListResponse {
        let encoded = category.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? category
        return try await request(path: "/practice/questions?category=\(encoded)&limit=\(limit)&offset=\(offset)")
    }

    func seedPracticeQuestions() async throws -> ImportResultResponse {
        try await request(path: "/practice/questions/seed", method: "POST")
    }

    func submitPracticeAttempt(questionID: String, answer: String, elapsedSeconds: Int?) async throws -> PracticeAttemptResponse {
        try await request(
            path: "/practice/attempt",
            method: "POST",
            body: PracticeAttemptRequest(questionID: questionID, answer: answer, elapsedSeconds: elapsedSeconds)
        )
    }

    func listResumes(limit: Int = 20, offset: Int = 0) async throws -> [ResumeRecord] {
        try await request(path: "/resumes?limit=\(limit)&offset=\(offset)")
    }

    func importResume(filename: String, text: String) async throws -> ResumeRecord {
        let encoded = Data(text.utf8).base64EncodedString()
        return try await request(
            path: "/resumes",
            method: "POST",
            body: ResumeImportRequest(filename: filename, contentBase64: encoded, sourcePath: nil)
        )
    }

    func deleteResume(id: String) async throws -> DeleteResponse {
        try await request(path: "/resumes/\(id)", method: "DELETE")
    }

    func createSession(_ body: CreateSessionRequest) async throws -> ChatResponse {
        try await request(path: "/sessions", method: "POST", body: body)
    }

    func listSessions(limit: Int = 20, offset: Int = 0) async throws -> [SessionSummary] {
        try await request(path: "/sessions?limit=\(limit)&offset=\(offset)")
    }

    func getSession(id: String) async throws -> SessionDetail {
        try await request(path: "/sessions/\(id)")
    }

    func deleteSession(id: String) async throws -> DeleteResponse {
        try await request(path: "/sessions/\(id)", method: "DELETE")
    }

    func sendMessage(sessionID: String, message: String) async throws -> ChatResponse {
        try await request(
            path: "/sessions/\(sessionID)/messages",
            method: "POST",
            body: ["message": message]
        )
    }

    func streamMessage(sessionID: String, message: String) async throws -> [StreamEvent] {
        let text: String = try await rawText(
            path: "/sessions/\(sessionID)/stream",
            method: "POST",
            body: ["message": message]
        )
        return parseSSE(text)
    }

    private func request<Response: Decodable, Body: Encodable>(
        path: String,
        method: String = "GET",
        body: Body? = Optional<String>.none,
        headers: [String: String] = [:]
    ) async throws -> Response {
        guard let url = URL(string: path, relativeTo: baseURL)?.absoluteURL else {
            throw InterviewApiError.invalidResponse
        }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        headers.forEach { request.setValue($0.value, forHTTPHeaderField: $0.key) }
        setClientHeaders(on: &request)
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let body {
            request.httpBody = try encoder.encode(body)
        }

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw InterviewApiError.invalidResponse
        }
        guard (200..<300).contains(http.statusCode) else {
            let detail = try? decoder.decode(ErrorResponse.self, from: data)
            throw InterviewApiError.server(detail?.detail ?? "HTTP \(http.statusCode)")
        }
        return try decoder.decode(Response.self, from: data)
    }

    private func rawText<Body: Encodable>(
        path: String,
        method: String,
        body: Body
    ) async throws -> String {
        guard let url = URL(string: path, relativeTo: baseURL)?.absoluteURL else {
            throw InterviewApiError.invalidResponse
        }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        setClientHeaders(on: &request)
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = try encoder.encode(body)
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw InterviewApiError.invalidResponse
        }
        guard (200..<300).contains(http.statusCode) else {
            throw InterviewApiError.server("HTTP \(http.statusCode)")
        }
        return String(data: data, encoding: .utf8) ?? ""
    }

    private func setClientHeaders(on request: inout URLRequest) {
        let requestID = UUID().uuidString
        request.setValue(requestID, forHTTPHeaderField: "X-Request-ID")
        request.setValue(requestID, forHTTPHeaderField: "X-Client-Request-Id")
        request.setValue("ios", forHTTPHeaderField: "X-Client-Platform")
        request.setValue(clientVersion, forHTTPHeaderField: "X-Client-Version")
    }

    private func parseSSE(_ text: String) -> [StreamEvent] {
        text.components(separatedBy: "\n\n").compactMap { block in
            let lines = block.split(separator: "\n").map(String.init)
            guard !lines.isEmpty else { return nil }
            let event = lines.first { $0.hasPrefix("event:") }?
                .replacingOccurrences(of: "event:", with: "")
                .trimmingCharacters(in: .whitespacesAndNewlines) ?? "message"
            let dataText = lines.first { $0.hasPrefix("data:") }?
                .replacingOccurrences(of: "data:", with: "")
                .trimmingCharacters(in: .whitespacesAndNewlines) ?? "{}"
            let data = (try? JSONSerialization.jsonObject(with: Data(dataText.utf8))) as? [String: Any] ?? [:]
            return StreamEvent(event: event, data: data)
        }
    }
}

private struct ErrorResponse: Codable {
    let detail: String?
}

enum InterviewApiError: LocalizedError {
    case invalidResponse
    case server(String)

    var errorDescription: String? {
        switch self {
        case .invalidResponse:
            return "服务响应无效"
        case .server(let message):
            return message
        }
    }
}
